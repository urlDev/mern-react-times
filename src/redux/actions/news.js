import axios from 'axios';

export const FETCH_POPULAR_SUCCESS = 'FETCH_POPULAR_SUCCESS';
export const FETCH_STORY_SUCCESS = 'FETCH_STORY_SUCCESS';
export const FETCH_NEWS_ERROR = 'FETCH_NEWS_ERROR';
export const CHANGE_HEADER = 'CHANGE_HEADER';

export const fetchPopularSuccess = (news) => ({
    type: FETCH_POPULAR_SUCCESS,
    payload: news,
});

export const fetchStorySuccess = (news) => ({
    type: FETCH_STORY_SUCCESS,
    payload: news,
});

export const changeHeader = (header) => ({
    type: CHANGE_HEADER,
    payload: header,
});

export const fetchNewsError = (error) => ({
    type: FETCH_NEWS_ERROR,
    payload: error,
});

export const fetchTopStories = (topStories) => async(dispatch) => {
    try {
        const response = await axios.get(
            `https://api.nytimes.com/svc/topstories/v2/${topStories}.json?api-key=${process.env.REACT_APP_NEWS_API_KEY}`,
        );
        const data = await response.data;
        return dispatch(fetchStorySuccess(data));
    } catch (error) {
        return dispatch(fetchNewsError(error));
    }
};

export const fetchMostPopular = () => async(dispatch) => {
    try {
        const response = await axios.get(
            `https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=${process.env.REACT_APP_NEWS_API_KEY}`,
        );
        const data = await response.data;
        return dispatch(fetchPopularSuccess(data));
    } catch (error) {
        return dispatch(fetchNewsError(error));
    }
};