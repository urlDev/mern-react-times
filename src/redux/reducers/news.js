import {
  FETCH_NEWS_ERROR,
  FETCH_POPULAR_SUCCESS,
  FETCH_STORY_SUCCESS,
  CHANGE_HEADER,
  CLEAN_STATE,
  CLEAR_ERROR,
  SET_WIDTH,
  OPEN_RESPONSIVE_MENU,
  CLOSE_RESPONSIVE_MENU,
} from "../actions/news";

const initialState = {
  loadingNews: true,
  popular: [],
  story: [],
  errorNews: null,
  header: "home",
  width: window.innerWidth,
  responsiveMenu: false,
};

const newsReducer = (state = initialState, action) => {
  switch (action.type) {
    case CLEAR_ERROR:
      return {
        ...state,
        errorNews: null,
      };
    case FETCH_POPULAR_SUCCESS:
      return {
        ...state,
        loadingNews: false,
        popular: action.payload.results.slice(0, 4),
      };
    case FETCH_STORY_SUCCESS:
      return {
        ...state,
        loadingNews: false,
        story: action.payload.results.slice(0, 3),
      };
    case FETCH_NEWS_ERROR:
      return {
        ...state,
        errorNews: action.payload,
      };
    case CHANGE_HEADER:
      return {
        ...state,
        header: action.payload,
      };
    case SET_WIDTH:
      return {
        ...state,
        width: action.payload,
      };
    case OPEN_RESPONSIVE_MENU:
      return {
        ...state,
        responsiveMenu: true,
      };
    case CLOSE_RESPONSIVE_MENU:
      return {
        ...state,
        responsiveMenu: false,
      };
    default:
      return state;
  }
};

export default newsReducer;
