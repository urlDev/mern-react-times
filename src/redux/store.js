import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));