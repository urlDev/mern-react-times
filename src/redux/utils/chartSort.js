const chartSort = (prevState, newState) => {
    const tempState = [...prevState, [...newState]];

    let sortedState = [...tempState];

    sortedState = sortedState
        .sort((a, b) => a[1] - b[1])
        .map((stock) => stock.slice(0, 1));

    return sortedState;
};

export default chartSort;