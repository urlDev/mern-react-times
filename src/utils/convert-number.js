const convertNumber = (labelValue) => {
  // Nine Zeroes for Billions
  return Math.abs(Number(labelValue)) >= 1.0e9
    ? (Math.abs(Number(labelValue)) / 1.0e9).toFixed(2) + "B"
    : // Six Zeroes for Millions
    Math.abs(Number(labelValue)) >= 1.0e6
    ? (Math.abs(Number(labelValue)) / 1.0e6).toFixed(2) + "M"
    : // Three Zeroes for Thousands
    Math.abs(Number(labelValue)) >= 1.0e3
    ? (Math.abs(Number(labelValue)) / 1.0e3).toFixed(2) + "K"
    : Math.abs(Number(labelValue));
};

export default convertNumber;
