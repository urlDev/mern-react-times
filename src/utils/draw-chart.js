// Reusable function to draw historical price chart in different components
const drawChart = ({ width, reDraw, chartData, svg, d3, index }) => {
  // reDraw => Charts in market details
  // because they are drawn according to screen sizes

  let height;
  let padding;
  let chartWidth;
  let dates;

  // reDraw is for when marketDetails page is shown
  // Chart is constantly redrawn then so thats why I made a separation
  if (reDraw) {
    // changing chart width according to screen width
    chartWidth = width < 1400 ? (width < 700 ? width - 30 : width - 200) : 550;
    height = width < 1400 ? chartWidth * 0.65 : 345;
    padding = 50;
  } else {
    // when its used in market components, width and height
    chartWidth = 100;
    height = 70;
    padding = 10;
  }

  svg.attr("width", chartWidth).attr("height", height);

  if (reDraw) {
    dates = chartData.map((value) => new Date(value.date));
  } else {
    // In market components, there are more than one chart
    // So I get their indexes from the component
    dates = chartData[index].map((value) => new Date(value.date));
  }

  // xScale
  const xAxisScale = d3
    .scaleTime()
    .domain([d3.min(dates), d3.max(dates)])
    .range(
      // Depending if its reDrawn or width changes,
      // In smaller screens, I don't put any padding
      reDraw
        ? width < 500
          ? [0, chartWidth]
          : [padding, chartWidth - padding]
        : [padding, chartWidth - padding]
    );

  // yScale
  const yAxisScale = d3
    .scaleLinear()
    .domain([
      // Chart is drawn from close data on each dataset
      // So in here I am checking where the function is being used
      d3.min(reDraw ? chartData : chartData[index], (data) => data.close),
      d3.max(reDraw ? chartData : chartData[index], (data) => data.close),
    ])
    .range([height - padding, padding]);

  // Charts in market cards don't use axes
  if (reDraw) {
    //   create x Axes
    const xAxes = d3
      .axisBottom(xAxisScale)
      .ticks(width < 700 ? d3.timeDay.every(10) : 5)
      .tickFormat((date) => d3.timeFormat("%b %d, %I:%M")(date));

    // draw x axes
    width > 500 &&
      svg
        .append("g")
        .call(xAxes)
        .attr("transform", `translate(0,${height - padding})`);

    // create y axes
    const yAxes = d3.axisLeft(yAxisScale).tickFormat((d) => `$ ${d}`);

    // draw y axes
    width > 500 &&
      svg.append("g").call(yAxes).attr("transform", `translate(${padding},0)`);
  }

  // create line
  const line = d3
    .line()
    .x((d) => xAxisScale(new Date(d.date)))
    .y((d) => yAxisScale(d.close));

  // draw historical price path
  const path = svg
    .append("path")
    .data(reDraw ? [chartData] : [chartData[index]])
    .style("fill", " none")
    .attr("stroke", reDraw ? "black" : "white")
    .attr("stroke-width", reDraw ? "1.5" : "1.2")
    .attr("d", line);

  // get the length of the path for line transition
  const pathLength = path.node().getTotalLength();

  // transition for the line
  const transitionPath = d3.transition().ease(d3.easeSin).duration(2000);

  // setting pathsLength for strokes attrs and using transition
  // I made it minus pathLength because it was drawing it backwards
  path
    .attr("stroke-dashoffset", -pathLength)
    .attr("stroke-dasharray", pathLength)
    .transition(transitionPath)
    .attr("stroke-dashoffset", 0);

  path.interrupt("stroke-dashoffset");
  path.interrupt("stroke-dasharray");
};

export default drawChart;
