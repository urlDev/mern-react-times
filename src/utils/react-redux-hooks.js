// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
