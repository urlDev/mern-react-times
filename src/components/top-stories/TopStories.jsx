import React from "react";
import moment from "moment";
import { useSelector } from "utils/react-redux-hooks";

import StoryTopic from "components/story-topic/StoryTopic";
import placeholder from "assets/newsPlaceholder.png";

import { Date } from "components/story-topic-headers/StoryTopicHeaders.styles";
import { SeeAll } from "components/most-popular/MostPopular.styles";
import {
  TopStoriesContainer,
  StoryContainer,
  Tag,
  Title,
  SubTitle,
  TagContainer,
} from "./TopStories.styles";

const TopStories = () => {
  const { story } = useSelector((state) => state.news);

  return (
    <div>
      <StoryTopic />
      <TopStoriesContainer>
        {story.slice(0, 3).map((data) => {
          const url = data?.multimedia?.[0]?.url;

          return (
            <StoryContainer key={data.title}>
              <img
                src={url ?? placeholder}
                alt="story"
                style={{ width: !url ? "400px" : "100%" }}
              />
              <TagContainer>
                <h1>
                  <Tag>{data.section}</Tag>
                </h1>
                <Date margin="10px" width="30%">
                  {moment(data.created_date).format("MMM DD, YYYY")}
                </Date>
              </TagContainer>

              <Title header={data.title.split(" ")}>{data.title}</Title>
              <SubTitle>
                {data.abstract.split(" ").slice(0, 20).join(" ")}...
              </SubTitle>
              <SeeAll href={data.url} target="_blank" rel="noopener">
                Continue Reading
              </SeeAll>
            </StoryContainer>
          );
        })}
      </TopStoriesContainer>
    </div>
  );
};

export default TopStories;
