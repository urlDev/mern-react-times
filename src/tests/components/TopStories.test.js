import { mount, shallow } from "enzyme";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import { story } from "tests/fixtures/news";

import TopStories from "components/top-stories/TopStories";

describe("Testing TopStories component", () => {
  let wrapper;
  let store;

  beforeEach(() => {
    store = mockStore({
      width: 1000,
      story: story.results[0],
    });

    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());

    wrapper = mount(<TopStories store={store} />);
  });

  test("Should render component", () => {
    expect(wrapper).toHaveLength(1);
    // checking for if h1 has text of story section ('us' in here)
    expect(wrapper.find("span").first().text()).toEqual("us");
  });
  test("Should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Should show other url when width is bigger than 1200", () => {
    store = mockStore({
      width: 1300,
      story: story.results[0],
    });

    wrapper = shallow(<TopStories store={store} />);

    expect(wrapper).toHaveLength(1);
    // checking for if the img has a src prop still after changing width
    expect(wrapper.find("img").prop("src")).toBeTruthy();
  });
});
