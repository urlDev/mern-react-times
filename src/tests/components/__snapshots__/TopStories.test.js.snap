// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TopStories component Should match snapshot 1`] = `
.c0 {
  margin-top: 15px;
}

.c0 h1 {
  font-family: var(--font-header);
  font-size: var(--size-category-selected);
  font-weight: 900;
  margin: 0;
  text-transform: capitalize;
}

.c0 img {
  height: 100%;
}

.c6 {
  margin: 0;
  margin-left: 10px;
  width: 30%;
  font-family: var(--font-text);
  font-size: var(--size-sub-menu);
}

.c6 span {
  font-family: var(--font-header);
  font-weight: 900;
}

.c11 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #707070;
  font-size: var(--size-sub-menu);
  font-family: var(--font-links);
  font-weight: bold;
  margin-bottom: 20px;
  cursor: pointer;
  display: inline-block;
  position: relative;
  -webkit-transition: color 0.3s ease-in;
  transition: color 0.3s ease-in;
}

.c11:hover {
  color: black;
}

.c11:after {
  content: "";
  position: absolute;
  bottom: -7px;
  left: 0;
  width: 110px;
  background: var(--gray);
  height: 2px;
}

.c1 {
  display: grid;
  grid-template-columns: repeat(2,1fr);
  width: 100%;
  margin-top: 20px;
}

.c10 {
  font-family: var(--font-text);
  font-size: var(--size-sub-menu);
  color: var(--gray);
  margin: 10px 0;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 10px;
}

.c8 {
  font-family: var(--font-header);
  font-size: var(--size-header-long);
}

.c2 {
  border-right: 1.3px solid lightgray;
}

.c2:not(:first-child) {
  padding: 0 40px;
}

.c2:not(:first-child) img {
  display: none;
}

.c2:not(:first-child) .c9 {
  display: none;
}

.c2:not(:first-child) .c3 {
  padding-top: 0;
}

.c2:not(:first-child) .c7 {
  font-size: var(--size-category-selected);
}

.c2:first-child {
  grid-row: span 2;
  padding-right: 40px;
}

.c2:last-child {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-bottom: 7px;
  margin-top: 15%;
}

.c2:nth-child(2) {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-bottom: 7px;
}

.c5 {
  background: black;
  color: white;
  padding: 7px 15px;
  text-transform: uppercase;
  font-family: var(--font-links);
  font-size: var(--size-text);
  font-weight: 400;
}

@media (max-width:768px) {
  .c6 {
    display: none;
  }
}

@media (min-width:1400px) {
  .c1 {
    width: 1000px;
  }
}

@media (max-width:1400px) {
  .c1 {
    grid-template-columns: 1fr;
  }
}

@media (max-width:830px) {
  .c1 {
    margin-bottom: 50px;
  }
}

@media (max-width:1400px) {
  .c2:not(:first-child) {
    padding: 0 40px 0 0;
    margin-top: 30px;
  }

  .c2 img {
    width: 100%;
    object-fit: cover;
    object-position: center;
  }
}

@media (max-width:830px) {
  .c2 {
    border-right: none;
    position: relative;
  }

  .c2:after {
    content: "";
    position: absolute;
    left: 0;
    bottom: 0;
    width: 100%;
    height: 1.3px;
    background: lightgray;
  }

  .c2:not(:first-child) {
    padding: 20px 0;
    margin-top: 0;
  }

  .c2:first-child {
    padding-bottom: 20px;
    padding-right: 0;
  }

  .c2:last-child {
    padding: 20px 0;
    margin-top: 0;
  }
}

<TopStories
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <div>
    <StoryTopic>
      <styled.div>
        <div
          className="c0"
        >
          <h1 />
        </div>
      </styled.div>
    </StoryTopic>
    <styled.div>
      <div
        className="c1"
      >
        <styled.div
          key="As Christmas Nears, Virus Experts Look for Lessons From Thanksgiving and more"
        >
          <div
            className="c2"
          >
            <img
              alt="story"
              src="https://static01.nyt.com/images/2020/12/19/us/covid-thanksgiving-effect-promo-1608423451574/covid-thanksgiving-effect-promo-1608423451574-thumbLarge.jpg"
              style={
                Object {
                  "width": "100%",
                }
              }
            />
            <styled.div>
              <div
                className="c3 c4"
              >
                <h1>
                  <styled.span>
                    <span
                      className="c5"
                    >
                      us
                    </span>
                  </styled.span>
                </h1>
                <styled.p
                  margin="10px"
                  width="30%"
                >
                  <p
                    className="c6"
                    width="30%"
                  >
                    Dec 20, 2020
                  </p>
                </styled.p>
              </div>
            </styled.div>
            <styled.h1
              header={
                Array [
                  "As",
                  "Christmas",
                  "Nears,",
                  "Virus",
                  "Experts",
                  "Look",
                  "for",
                  "Lessons",
                  "From",
                  "Thanksgiving",
                  "and",
                  "more",
                ]
              }
            >
              <h1
                className="c7 c8"
              >
                As Christmas Nears, Virus Experts Look for Lessons From Thanksgiving and more
              </h1>
            </styled.h1>
            <styled.h3>
              <h3
                className="c9 c10"
              >
                Public health experts warned that large holiday gatherings would spread the coronavirus broadly. Weeks later, the data is complicated.
                ...
              </h3>
            </styled.h3>
            <styled.a
              href="https://www.nytimes.com/interactive/2020/12/20/us/covid-thanksgiving-effect.html"
              rel="noopener"
              target="_blank"
            >
              <a
                className="c11"
                href="https://www.nytimes.com/interactive/2020/12/20/us/covid-thanksgiving-effect.html"
                rel="noopener"
                target="_blank"
              >
                Continue Reading
              </a>
            </styled.a>
          </div>
        </styled.div>
      </div>
    </styled.div>
  </div>
</TopStories>
`;
